crates/rustc_codegen_spirv/src/builder/load_store.rs

```diff
@@ -0,0 +1,506 @@
+use rustc_middle::bug;
+
+use super::Builder;
+use crate::builder_spirv::{BuilderCursor, SpirvValue, SpirvValueExt};
+use crate::codegen_cx::BindlessDescriptorSets;
+use crate::rustc_codegen_ssa::traits::BuilderMethods;
+use crate::spirv_type::SpirvType;
+use rspirv::spirv::Word;
+use rustc_span::DUMMY_SP;
+use rustc_target::abi::{Abi, Align, Scalar, Size};
+
+impl<'a, 'tcx> Builder<'a, 'tcx> {
+    // walk down every member in the ADT recursively and load their values as uints
+    // this will break up larger data types into uint sized sections, for
+    // each load, this also has an offset in dwords.
+    fn recurse_adt_for_stores(
+        &mut self,
+        uint_ty: u32,
+        val: SpirvValue,
+        base_offset: u32,
+        uint_values_and_offsets: &mut Vec<(u32, SpirvValue)>,
+    ) {
+        let ty = self.lookup_type(val.ty);
+
+        match ty {
+            SpirvType::Adt {
+                ref field_types,
+                ref field_offsets,
+                ref field_names,
+                ..
+            } => {
+                for (element_idx, (ty, offset)) in
+                    field_types.iter().zip(field_offsets.iter()).enumerate()
+                {
+                    let load_res = self.extract_value(val, element_idx as u64);
+
+                    if offset.bytes() as u32 % 4 != 0 {
+                        let adt_name = self.type_cache.lookup_name(val.ty);
+                        let field_name = if let Some(field_names) = field_names {
+                            &field_names[element_idx]
+                        } else {
+                            "<unknown>"
+                        };
+
+                        self.fatal(&format!(
+                            "Trying to store to unaligned field: `{}::{}`. Field must be aligned to multiple of 4 bytes, but has offset {}",
+                            adt_name,
+                            field_name,
+                            offset.bytes() as u32));
```

Comment:
Is there a reason this is fatal and not an error? Seems better to make it an error.

Reply:
Switched it